staging/src/k8s.io/kubectl/pkg/cmd/debug/debug.go

```diff
@@ -0,0 +1,454 @@
+	r := o.builder.
+		WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+		NamespaceParam(o.Namespace).DefaultNamespace().ResourceNames("pods", o.PodNames...).
+		Do()
+	if err := r.Err(); err != nil {
+		return fmt.Errorf("error creating builder: %v", err)
+	}
+
+	ctx := context.Background()
+	err := r.Visit(func(info *resource.Info, err error) error {
+		if err != nil {
+			// TODO(verb): configurable early return
+			return err
+		}
+
+		pods := o.podClient.Pods(info.Namespace)
+		ec, err := pods.GetEphemeralContainers(ctx, info.Name, metav1.GetOptions{})
+		if err != nil {
+			// The pod has already been fetched at this point, so a NotFound error indicates the ephemeralcontainers subresource wasn't found.
+			if serr, ok := err.(*errors.StatusError); ok && serr.Status().Reason == metav1.StatusReasonNotFound {
+				return fmt.Errorf(errDisabled, err)
+			}
+			return err
+		}
+		klog.V(2).Infof("existing ephemeral containers: %v", ec.EphemeralContainers)
+
+		debugContainer := o.generateDebugContainer(info.Object.(*corev1.Pod))
+		klog.V(2).Infof("new ephemeral container: %#v", debugContainer)
+		ec.EphemeralContainers = append(ec.EphemeralContainers, *debugContainer)
+		_, err = pods.UpdateEphemeralContainers(ctx, info.Name, ec, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("updating ephemeral containers: %v", err)
+		}
+
+		if o.Attach {
+			opts := &attach.AttachOptions{
+				StreamOptions: exec.StreamOptions{
+					IOStreams: o.IOStreams,
+					Stdin:     o.Interactive,
+					TTY:       o.TTY,
+					Quiet:     o.Quiet,
+				},
+				CommandName: cmd.Parent().CommandPath() + " attach",
+
+				Attach: &attach.DefaultRemoteAttach{},
+			}
+			config, err := f.ToRESTConfig()
+			if err != nil {
+				return err
+			}
+			opts.Config = config
+			opts.AttachFunc = attach.DefaultAttachFunc
+
+			attachablePod, err := polymorphichelpers.AttachablePodForObjectFn(f, info.Object, opts.GetPodTimeout)
+			if err != nil {
+				return err
+			}
+			err = handleAttachPod(ctx, f, o.podClient, attachablePod.Namespace, attachablePod.Name, debugContainer.Name, opts)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+func containerNames(pod *corev1.Pod) map[string]bool {
+	names := map[string]bool{}
+	for _, c := range pod.Spec.Containers {
+		names[c.Name] = true
+	}
+	for _, c := range pod.Spec.InitContainers {
+		names[c.Name] = true
+	}
+	for _, c := range pod.Spec.EphemeralContainers {
+		names[c.Name] = true
+	}
+	return names
+}
+
+// generateDebugContainer returns an EphemeralContainer suitable for use as a debug container
+// in the given pod.
+func (o *Options) generateDebugContainer(pod *corev1.Pod) *corev1.EphemeralContainer {
+	name := o.Container
+	if name == "" {
+		cn, existing := "", containerNames(pod)
+		for cn == "" || existing[cn] {
+			cn = fmt.Sprintf("debugger-%s", nameSuffixFunc(5))
+		}
+		if o.ErrOut != nil {
```

Comment:
I can't imagine this not being true, so it's ok to leave this check out, every consumer is expected to inject std and err streams.